platform-sdk/docs/proposals/001-tss-rosters-design-proposal/proposal.md

```diff
@@ -0,0 +1,290 @@
+
+there should be absolutely no need to introduce any additional mechanisms or interfaces between Platform and Services to
+supply a genesis roster.
+
+If we must, we may add a field to the `config.txt` to indicate the number of shares for each node.
+This should be a relatively trivial change.
+However, since the `config.txt` would only ever be used once during genesis, it’s a lot simpler to skip that and assume
+equal shares for all genesis nodes, and only start supporting non-equal shares when a new Candidate Roster is installed
+via the new public interface of the Platform API. But as mentioned, if we must support this at genesis for brand new
+networks, we can do that, and it should be trivial.
+
+### Roster changes needed for Components
+
+- Reconnect. Reconnect logic currently exchanges and validates address book between the learner and teacher nodes. The
+  learner node uses the address book to select the teacher node, as well as remove all invalid signatures from the
+  state. Both of these will need to be updated to use rosters instead.
+- Networking. Most parts of the network code have been abstracted away from the address book as part of the Dynamic
+  Address Book effort, so there should be minimal work left there.
+- Event Validation. The event validation logic uses the address book to determine whether a given event has a valid
+  signature. This will be updated to use rosters instead.
+- Miscellaneous components. Other components that reference things stored in the address book (like consensus needing
+  the node weight or the crypto module verifying keys) will need to be updated to use things stored in rosters instead.
+
+### Test Plan
+
+Some of the obvious test cases to be covered in the plan include validating one or more of the following scenarios:
+
+1. New valid Candidate Roster created with no subsequent one sent by app. Verify accept.
+2. New valid Candidate Roster created with subsequent one sent by app. Verify accept.
+3. Invalid roster(s) sent. Verify reject.
+4. Empty roster(s) sent. Verify reject.
+5. Node Failures During Roster Change: What happens if nodes fails or disconnects during a roster change? Verify valid
+   node successfully reconnects.
+6. Concurrent Roster Updates: What if we make multiple roster updates concurrently? Verify no effect on adoption.
+7. Rosters with nodes with zero or maximum shares. What happens?
+8. Roster recovery? Node receives roster, crashes. Wakes up … ?
+9. What testing do we need for genesis new network?
+10. What testing do we need for an already new network with no previous TSS upgrade
+11. What testing do we need for an already new network with previous TSS upgrade
+
+### Metrics
+
+We propose that some metrics be added to track the progress of the TSS key material and votes accumulation in the
+Candidate Roster.
+Others may include create timestamp of a Candidate Roster, and how long voting took.
+
+### Implementation and Delivery Plan
```

Comment:
How does this impact Services?  
How does this impact DevOps? 
Is a HIP required?

Reply:
DevOps Impact: 
* When adding new nodes to existing networks, they need to be given a state after the network has upgraded.  The state contains the address book /roister with the new node in it.    Devops will no longer use config.txt on new nodes to existing networks.

Reply:
who provides the state to devops, services? Should we include that here?

Reply:
I think we should indicate at a high level how Services and DevOps will need to change their processes and why.  

DevOps gets the state from disk when a snapshot is created.

Reply:
With Block Node devops will get state from a Block Node, and any require filtering can be done there.
Not sure if it's directly relevant, but perhaps something worth considering.

Reply:
I would hope that TSS-Roster is implemented and shipped before block node.  if not, we should adjust our language in this proposal to simply talk about the change in work flow in providing a state and not where the state comes from.    The state must have the node in its active roster.

Reply:
As per the project plan, TSS Roster delivery has to come before the rest of TSS spec.